### ISSUE TITLE
[Bug] Mamba radix cache can turn fresh prefill prefix hits into 0-hit after split

### ISSUE THREAD

**Issue by ChangyiYang** [sgl-project/sglang#22935]
### Checklist

- [x] This is a bug report, not a general question.
- [x] This report is in English.

### Summary

`MambaRadixCache` can degrade a structurally valid prefix match into an effective 0-hit on fresh prefills.

The problematic combination is:

1. fresh prefill lookup intentionally matches at most `input_len - 1` tokens in `Req.init_next_round_input()` so that one token is left for the current extend/logprob step,
2. if that lookup ends in the middle of a compressed radix edge, `_split_node()` creates a new internal prefix node with `mamba_value = None`,
3. `_match_prefix_helper()` advances `best_value_len` only when the current node has a `mamba_value`, and
4. `_match_post_processor()` truncates the returned KV indices to `value[:best_value_len]`.

Net effect: the tree can structurally match a prefix, but the returned `device_indices` can still become length 0.

### Why this seems surprising

The `input_len - 1` behavior itself makes sense for generation: if the entire input were prefix-hit, the current extend batch would have 0 tokens and would not produce the next-token logits.

However, in the Mamba radix cache, that legitimate `n - 1` lookup interacts badly with split semantics. A fresh request with an identical prompt can fail to reuse any device KV, even though the prompt prefix is already present in the radix tree.

### Minimal scenario

Assume request 1 has prompt `[A, B, C]`.

After request 1 prefill completes, `cache_unfinished_req()` inserts the current `req.fill_ids` into `MambaRadixCache`. In the common fresh-prefill case this is the prompt itself, e.g. `[A, B, C]`, with a valid `mamba_value` on that node.

Then request 2 arrives with the same prompt `[A, B, C]`.

During `Req.init_next_round_input()`:

- `fill_ids = origin_input_ids + output_ids = [A, B, C]`
- `max_prefix_len = input_len - 1 = 2`
- the cache lookup key becomes `[A, B]`, not `[A, B, C]`

If the tree currently contains a compressed child/key `[A, B, C]`, matching `[A, B]` causes a split:

- original: `root -> [A,B,C] (has mamba)`
- after split: `root -> [A,B] (mamba=None) -> [C] (original suffix)`

The fresh request stops at `[A,B]` because that is the entire lookup key for this round.

At that point:

- the path was structurally matched,
- but the matched node has `mamba_value = None`,
- so `best_value_len` does not advance,
- and `_match_post_processor()` truncates the returned KV hit to length 0.

### Code paths involved

- `python/sglang/srt/managers/schedule_batch.py`
  - `Req.init_next_round_input()`
  - comment: "the matched length is at most 1 less than the input length to enable logprob computation"
- `python/sglang/srt/mem_cache/mamba_radix_cache.py`
  - `_split_node()` sets `new_node.mamba_value = None`
  - `_match_prefix_helper()` updates `best_value_len` only when `node.mamba_value is not None`
  - `_match_post_processor()` truncates `value` to `value[:best_value_len]`

### Impact

This can make prompt reuse much worse than expected for sequential workloads and rollout-style workloads:

- identical prompt reuse across fresh requests can collapse to 0-hit,
- shared-prefix reuse is much more brittle than in the standard radix cache,
- the behavior is especially easy to trigger because fresh prefill intentionally uses `n - 1` lookup.

### Expected behavior

At minimum, it would be helpful to confirm whether this is intended.

If not intended, possible fixes could include one of the following directions:

- preserve usable prefix reuse across split boundaries for fresh prefill lookups,
- avoid truncating full KV hits solely because the split/internal node has no `mamba_value`, and rely on `mamba_branching_seqlen` to re-track state as needed,
- or introduce explicit intermediate checkpointing so the `n - 1` fresh-prefill lookup lands on a node with a valid Mamba state more often.

### Related issue

This seems related to, but distinct from, #22326:

- #22326 is a feature proposal to add more Mamba checkpoints during prefill,
- this report is about current behavior where a valid fresh-prefill prefix match can degrade to 0-hit because of split semantics plus Mamba-gated KV truncation.

**Comment by s5u13b** [comment]
Hello, I'm running into the same issue as described above. Could anyone share if there has been any progress or a known workaround? Thank you in advance.

**Comment by rouchenzi** [comment]
Observed the same issue on qwen 3.5, it seems conditional on prompt length vs chunked_prefill_size, breaking prefix caching for short prompts but might have less impact on long-context workloads. 

For short prompts (< chunked_prefill_size) saw 0% hit as described above that the first request inserts a node, the second request's n-1 match splits it into a mamba=None node that persists and blocks all future matches. For longer prompts seems the first request triggers chunked prefill, which inserts nodes with valid mamba_value every chunked_prefill_size tokens. and subsequent requests can fully match through these nodes without triggering split, only the last matched node gets split due to n-1.

Tested command(SGLang 0.5.9, chunked_prefill_size=8192)
```
# 4K system prompt (prefill batch cached_tokens = 0):
python3 -m sglang.bench_serving --backend sglang --dataset-name generated-shared-prefix \
    --gsp-system-prompt-len 4000 --gsp-question-len 128 --gsp-output-len 16 \
    --gsp-num-groups 5 --gsp-prompts-per-group 20 --num-prompts 100 --max-concurrency 1

# 32K system prompt (prefill batch cached tokens > 0, most tokens cached):
python3 -m sglang.bench_serving --backend sglang --dataset-name generated-shared-prefix \
    --gsp-system-prompt-len 32000 --gsp-question-len 256 --gsp-output-len 16 \
    --gsp-num-groups 5 --gsp-prompts-per-group 20 --num-prompts 100 --max-concurrency 1
```

**Comment by hanming-lu** [comment]
Try mamba scheduler strategy extra_buffer

**Comment by waynel96** [comment]
Tested `extra_buffer` on this — it fixes the identical-prompt case and short-suffix extends, but not the shared-prefix + long-suffix case (the common multi-sample / RL-rollout pattern). Sharing measured numbers in case they help.

**Setup:** SGLang v0.5.9, Qwen3.5-122B-A10B, TP8, `--chunked-prefill-size -1` (chunked prefill **off**, so only the `extra_buffer` end-checkpoint is in play), `--page-size 64`. The cache holds only `P+s1` (P = 6476 tokens, a shared prefix), then I query `P+s2` — same `P`, a different suffix `s2` diverging at the first suffix token. 

| suffix length | `no_buffer` (V1) | `extra_buffer` (V2) |
|---|---|---|
| 40 tokens (`< 64`) | 0.000 | **0.992** |
| 376 tokens | 0.000 | **0.000** |
| 1296 tokens | 0.000 | **0.000** |

From my investigation, it seems `extra_buffer` inserts one valid-`mamba_value` checkpoint per request at `mamba_last_track_seqlen = floor(len/64)×64` (64 = `mamba_cache_chunk_size`, which falls back to `FLA_CHUNK_SIZE` here since the model has no `mamba_chunk_size`). That checkpoint lands **inside the shared `P`** only when the suffix is short enough to be floored away (`< 64` tokens for a 64-aligned `P`). For a realistic suffix (`≥ 64`), the checkpoint floors *into the suffix*, past the point where `P+s2` diverges — so the lookup still splits mid-edge onto the `mamba_value=None` node, exactly as in the original report. 


**Question:** `extra_buffer` clearly handles the re-sent / shared system-prompt case well (the `floor/64` checkpoint aligns and hits). For the **shared-prefix + divergent long-suffix** pattern above, though, the single checkpoint floors past the divergence point, so it still 0-hits. Is supporting that case on the radar (e.g. finer-grained prefill checkpointing, along the lines discussed in #22326), or is the single end-of-prefill checkpoint the intended design for now? Either way is useful to know — just trying to understand the intended scope.

**Comment by Destination2020** [comment]
I did a source-level and deterministic-probe pass on the current cache paths and would like to align on the intended design before attempting an implementation.

**Baseline.** The current `main` I rechecked is [`24d625698d44c78f6e8ab8b7c19f96f45bbaa90a`](https://github.com/sgl-project/sglang/commit/24d625698d44c78f6e8ab8b7c19f96f45bbaa90a). The probes and selected upstream tests were executed at `880ab72f34ead8f39544c7c7e08edbd443964a1f`; the two commits between that pin and current `main` do not touch the radix/Mamba/scheduler paths.

**Legacy vs Unified.** The legacy `MambaRadixCache` reproduces the exact report: matching `[A,B]` against cached `[A,B,C]` splits at depth 2, the new parent has no `mamba_value`, and the safe returned length is 0. The default Unified path preserves more information: it records the deeper Full-Attention match in `full_kv_hit_length` and derives a future Mamba branch-checkpoint target, while `device_indices` remains at the all-component-safe boundary. So Unified avoids silently treating an unavailable Mamba state as ready, but the first divergent request can still recompute despite a deeper structural/Full-KV match.

The deterministic cases were:

| Case | Structural / Full-Attention progress | Exact Mamba checkpoint | Safe scheduler hit |
|---|---:|---:|---:|
| Identical `[A,B,C]`, fresh lookup capped to `[A,B]` | 2 | 0 | 0 initially; 2 after recomputation installs the branch checkpoint |
| Short divergence with a retained checkpoint at the shared boundary | 4 | 4 | 4 |
| Long divergence where the previous checkpoint lies inside the old suffix | 4 | 0 | 0 initially, with branch target 4 |

This is consistent with the measured `extra_buffer` behavior already reported above: identical/short-suffix reuse improves, while a first long-suffix divergence can still have no checkpoint at the shared-prefix boundary.

I do not think returning `full_kv_hit_length` directly as the global cached length is correct. A Mamba state is a point value at an exact token depth, not a downward-closed prefix: state at the old leaf cannot be used at the new split parent. Also, with interleaved Attention/Mamba layers, starting forward at the deeper Attention boundary while the recurrent state is shallower cannot generally be implemented as “recompute only Mamba.” Without a deeper exact state provider or a separate replay executor, the current common safe boundary is the correct fallback. This is also the general safety pattern in vLLM's recent hybrid partial-hit work ([vllm#51468](https://github.com/vllm-project/vllm/pull/51468), [vllm#51843](https://github.com/vllm-project/vllm/pull/51843)): divergent progress is retained only when the missing state can actually be supplied; otherwise it falls back to a common boundary.

**Possible bounded-checkpoint direction.** My current preference would be to keep adaptive branch checkpoints and optionally add a conservative, default-off interval policy:

- checkpoint only at exact `lcm(tree_page_size, backend_state/checkpoint_grid)` depths;
- keep turn-end/leaf and learned branch checkpoints;
- optionally retain interval checkpoints, capped by `max_states_per_path` and explicit device/host byte budgets;
- let Mamba checkpoints be evicted independently without deleting valid Full KV;
- require backend/HiCache/PD/spec/session capability declarations and fail closed for unknown combinations;
- only consider enabling it after cold-vs-warm output/logprob equality and TTFT/throughput measurements.

Before caps, interval checkpoint state memory and extra linear radix nodes are both approximately `O(path_length / interval)`; the state bytes are approximately `retained_checkpoint_count × bytes_per_Mamba_state`, with corresponding host/L3 transfer growth if those checkpoints are offloaded. Hard per-path and byte caps are therefore part of the correctness/operability contract, not a later optimization.

Would maintainers prefer:

1. keeping only the current adaptive branch-checkpoint policy and hardening/clarifying its progress and lifetime contract; or
2. exploring adaptive checkpoints plus a capped, default-off interval policy for the first long-divergent-suffix hit?

If (2) is in scope, I can first prepare a small design/prototype with exact progress invariants, deterministic split tests, checkpoint/node/memory accounting, and one CUDA Hybrid-Mamba correctness/performance comparison before proposing production changes.

**Comment by Zhang-Wen-chao** [comment]
A fix for the default-strategy case is up for review: #38625.

Root-cause recap, matching the analysis upthread: the n-1 prefix lookup issued by design (one token reserved for logprob computation) lands mid-edge, so the match splits the node — and the split interior node carries no mamba checkpoint. Since a mamba state is a point value at the node's end depth and cannot be truncated backwards, the match safely truncates to 0. In the default `no_buffer` strategy, checkpoints are only created at insert depth (request end, or chunk ends under chunked prefill), so the interior node is always bare and identical prompts stay at 0% reuse, exactly as reported here.

One thing worth flagging for anyone reproducing on current main: the default tree cache there is now `UnifiedRadixCache`, and the same semantics live in its MAMBA component (its insert path is documented as mirroring MambaRadixCache). The legacy `MambaRadixCache` class is no longer reachable from server startup, so the issue now reproduces through the unified path — the PR ports the fix to both.

The fix: during `no_buffer` prefill, persist the intermediate state at the deepest interior grid boundary (grid = lcm(page alignment, mamba chunk size)) using the same track machinery `extra_buffer` already uses, and insert it as an interior checkpointed node. One extra state per request — O(1) memory. `extra_buffer` (as suggested upthread) remains the right immediate workaround; this brings equivalent coverage to the default strategy without a strategy change.

Validation on a GDN-family hybrid (Qwen3.8-27B fp8, 4096-token prompts):

- identical prompts: `cached_tokens` 0 → 4032/4096 (4032 = ((4096-1)//64)*64, the interior boundary), e2e 1.49s → 0.35s
- shared-prefix requests forking at or beyond the deepest interior boundary reuse 4032 — a mid-edge fork no longer destroys the reuse
- greedy (temperature=0) outputs are byte-identical with and without cache hits, so the state resume is exact
- the `extra_buffer` path is bit-identical before/after the change
- CPU contract tests pin the split/tombstone semantics so this regression cannot come back silently

Honest boundaries: prompts shorter than `grid + 1` have no interior state to snapshot (kernel constraint); divergence earlier than the deepest interior boundary of the last extend still falls to 0 — that is the bounded interval-checkpointing direction @Destination2020 analyzed upthread, which this PR deliberately does not preclude. Speculative decoding, int8 checkpoints, and `extra_buffer` keep their current behavior.

The branch is `fix/mamba-n1-checkpoint` on my fork if anyone wants to try it on their workload. Would appreciate a review and a `run-ci` label from the maintainers when someone gets a chance.
